octopusdeploy/resource_project.go

```diff
@@ -81,13 +81,21 @@ func resourceProject() *schema.Resource {
 				Optional: true,
 				Default:  false,
 			},
-			"deployment_step_windows_service": getDeploymentStepWindowsServiceSchema(),
-			"deployment_step_iis_website":     getDeploymentStepIISWebsiteSchema(),
-			"deployment_step_inline_script":   getDeploymentStepInlineScriptSchema(),
-			"deployment_step_kubernetes_helm": getDeploymentStepKubernetesHelmSchema(),
-			"deployment_step_kubernetes_yaml": getDeploymentStepKubernetesYamlSchema(),
-			"deployment_step_package_script":  getDeploymentStepPackageScriptSchema(),
-			"deployment_step_apply_terraform": getDeploymentStepApplyTerraformSchema(),
+			"deployment_step": {
```

Comment:
Let's make this plural and encourage folks to put all the deployment steps in a single block as that will make for an easier read when looking at the configuration. For example:

this:
```
deployment_step {
  windows_service {
  ...
  }
}

deployment_step {
  inline_script {
  ...
  }
}
```

would become:

```
deployment_steps {
  windows_service {
  ...
  }

  inline_script {
  ...
  }
}
```

Reply:
Wouldn't this basically result in the exact same behaviour that we're trying to fix? I don't know if we should recommend this type of pattern if the order is not guaranteed to match what you write.

Reply:
The two examples are functionally equivalent. Since `deployment_steps` is declared as type List it will always preserve the order. The unit test that you added confirms this by checking that the steps are in the correct indexes in the list.

Reply:
@jeff-french For example, this:

```tf
deployment_step  {
  kubernetes_yaml {
    step_name = "Kubernetes YAML"
    step_condition = "success"
    yaml_values = <<EOF
# a bunch of yaml
EOF

      target_roles = [ "Linux" ]
  }
}

deployment_step {
  inline_script {
    step_name = "Run script"
    script_type = "Bash"
    script_body = <<EOF
#!/bin/bash

echo "Hello, World!"
EOF

    target_roles = [ "Linux" ]
  }
}
```

and this:

```tf
deployment_steps  {
  kubernetes_yaml {
    step_name = "Kubernetes YAML"
    step_condition = "success"
    yaml_values = <<EOF
# a bunch of yaml
EOF

      target_roles = [ "Linux" ]
  }

  inline_script {
    step_name = "Run script"
    script_type = "Bash"
    script_body = <<EOF
#!/bin/bash

echo "Hello, World!"
EOF

    target_roles = [ "Linux" ]
  }
}
```

results in a different order once actually applied into Octopus. This is because, for each element of the array of deployment_steps, it checks the keys (`kubernetes_yaml`, `inline_script`) in a specific order